### ISSUE TITLE
[Bug]: Collection keywords as list containing None

### ISSUE THREAD

**Issue by TLouf** [matplotlib/matplotlib#21179]
### Bug summary

Collection style keywords can be set to `None` to retrieve the default, but when passing them as a list containing a `None` , matplotlib raises an error.

### Code for reproduction

```python
from matplotlib.collections import Collection
coll = Collection(edgecolors=[None]) # or facecolors, linestyles or linewidths
```

### Actual outcome

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_9904/3018246391.py in <module>
----> 1 Collection(edgecolors=['r', None])

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/_api/deprecation.py in wrapper(*inner_args, **inner_kwargs)
    429                          else deprecation_addendum,
    430                 **kwargs)
--> 431         return func(*inner_args, **inner_kwargs)
    432 
    433     return wrapper

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/collections.py in __init__(self, edgecolors, facecolors, linewidths, linestyles, capstyle, joinstyle, antialiaseds, offsets, transOffset, norm, cmap, pickradius, hatch, urls, offset_position, zorder, **kwargs)
    177         self._hatch_color = mcolors.to_rgba(mpl.rcParams['hatch.color'])
    178         self.set_facecolor(facecolors)
--> 179         self.set_edgecolor(edgecolors)
    180         self.set_linewidth(linewidths)
    181         self.set_linestyle(linestyles)

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/collections.py in set_edgecolor(self, c)
    837             c = c.lower()
    838         self._original_edgecolor = c
--> 839         self._set_edgecolor(c)
    840 
    841     def set_alpha(self, alpha):

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/collections.py in _set_edgecolor(self, c)
    816             self.stale = True
    817             return
--> 818         self._edgecolors = mcolors.to_rgba_array(c, self._alpha)
    819         if set_hatch_color and len(self._edgecolors):
    820             self._hatch_color = tuple(self._edgecolors[0])

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/colors.py in to_rgba_array(c, alpha)
    381             rgba = np.array(c)
    382         else:
--> 383             rgba = np.array([to_rgba(cc) for cc in c])
    384     else:
    385         rgba = np.array([to_rgba(cc) for cc in c])

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/colors.py in <listcomp>(.0)
    381             rgba = np.array(c)
    382         else:
--> 383             rgba = np.array([to_rgba(cc) for cc in c])
    384     else:
    385         rgba = np.array([to_rgba(cc) for cc in c])

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/colors.py in to_rgba(c, alpha)
    204         rgba = None
    205     if rgba is None:  # Suppress exception chaining of cache lookup failure.
--> 206         rgba = _to_rgba_no_colorcycle(c, alpha)
    207         try:
    208             _colors_full_map.cache[c, alpha] = rgba

~/miniconda3/envs/geopandas-dev-tests/lib/python3.9/site-packages/matplotlib/colors.py in _to_rgba_no_colorcycle(c, alpha)
    282     # tuple color.
    283     if not np.iterable(c):
--> 284         raise ValueError(f"Invalid RGBA argument: {orig_c!r}")
    285     if len(c) not in [3, 4]:
    286         raise ValueError("RGBA sequence should have length 3 or 4")

ValueError: Invalid RGBA argument: None
```

### Expected outcome

We enter the logic contained within this `if` statement,

https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/collections.py#L773-L777

and the Collection is initiated with a list-like `edgecolor` with the None replaced with the default value. Same for `facecolors`, `linestyles` or `linewidths`. 

### Matplotlib Version

3.4.3

**Comment by timhoffm** [comment]
Thanks for the report. We are not claiming that lists containing *None* are supported right now. I'm labelling this as new feature, because it's a logical extension of the API.

**Comment by TLouf** [comment]
I think it can be achieved by casting the arguments to arrays in `_set_{kwarg}` and using boolean masks, but surely it's not as simple as it sounds as it will probably entail changes in other parts of the code. 

May I make an attempt at it and submit a PR if I manage to get something that passes tests?

**Comment by timhoffm** [comment]
Of course, you are very welcome to submit a PR!

I'd try to be very conservative on casting the inputs.

Checking the possible [color specifications](https://matplotlib.org/devdocs/tutorials/colors/colors.html), only RGB(A) tuples are sequence-like. While we accept any sequence-like type for RGB(A), in particular also lists ([1, 0, 0] is the same as (1, 0, 0) is the same as 'red'), these tuples cannot contain `None`. It should therefore be safe to do something like
```
if isinstance(colors, list):
    colors = [mpl.rcParams[...] if c is None else c for c in colors]
```

**Comment by mwaskom** [comment]
> Checking the possible color specifications, only RGB(A) tuples are sequence-like.

You likely know this, but just flagging that strings both quack like a sequence and inherit from `collections.abc.Sequence`. Distinguishing strings from "generic ordered collections of data" is an annoyingly tricky problem.

Please test this carefully! There's a history of seemingly innocuous changes to facecolor/edgecolor logic breaking downstream projects :)

**Comment by jklymak** [comment]
Yeah I'm pretty skeptical about this.  Bare None meaning "default" is (for better or worse) our convention.  None in a list falling back to the default makes a bit of sense but I'm not convinced it is worth the significant complication to the code.  If we do this there needs to be a super compelling case why this will make everyone's life better.

**Comment by tacaswell** [comment]
I am also relatively skeptical on this as it is not obvious to me what the "default" color should be in a list like this.  Given that scatter participates in the color cycle if you do not specify the color at all, would you expect to grab the next color once and use it for all of the `None` in the list or to grab a new color for each `None` in the list?

**Comment by TLouf** [comment]
I just opened a draft PR to show my proposed changes. 

> I am also relatively skeptical on this as it is not obvious to me what the "default" color should be in a list like this. Given that scatter participates in the color cycle if you do not specify the color at all, would you expect to grab the next color once and use it for all of the None in the list or to grab a new color for each None in the list?

Here I simply want to enable `None` for collections, without enabling it in `scatter` or any other higher level functions. As of now, in my PR, a call to these with lists containing `None` raises, as, for instance, a check is made on colors in `scatter` before passing it to a collection. The issue I'm trying to solve here is that for instance, in [GeoPandas](https://github.com/geopandas/geopandas/blob/master/geopandas/plotting.py#L126), plots are made by creating the right collections, and a user cannot pass one of these kwargs as an iterable containing a `None` to get the default, which is not always trivial to retrieve.

**Comment by tacaswell** [comment]
Ah, I see this is a layer down so there is no color cycle to deal with so it is unambigious to always use the same value as the filler.

That said, I'm still very skeptical of this proposal.  

As @jklymak  said above There is a very strong convention through Matplotlib (and Python!) that `None` passed for a parameter means "the user does not have anything to say about this value, do what you will" (and as a stand-in for having a default value in the signature for cases where you either want to make the default value run-time variable or the default value is a mutable object) in cases where `None` is not a valid value for the parameter (and you can play other games with sentinels if `None` _can_ be a valid value and you need to know if the user passed the keyword or not, but I digress).  However, extending that to the _contents_ of iterables is, to my knowledge, not a well established convention (I am not aware of any other libraries that extend their default handling to the inside of data structures passed in).  It also leads to some weird differences between what can be expressed in a numpy array vs a list/tuple (and as implemented the lw code changes the behavior of passing a linewidth of `np.nan` (which currently does not render, if that is the right behavior is a different discussion)).  All of this together makes documenting the input types to the keyword arguments harder to explain which the Zen of Python advises against.

Further, I am not sure of the use case for this.  If you want to pass only a few of the colors / linewidths, then presumably it is so that you can highlight those elements of the collection from the others.  However, because the default values here are controlled by rcParams, it is possible that the default has been changed to the "hi-light" value, but because everything else falls back to the same value you have lost the hi-light (where as if you had set all of the values you could be sure the hi-light would remain different).

Finally, one of the things that makes Matplotlib hard to use is that different parts of the library have slightly different conventions and the inconsistency can be maddening for users.  If we start down the path of filling `None`s inside of user supplied data structures, we probably should do that _everywhere_ where we take a {`None`, scalar, vector} is input.  I worry that if we go down this path we are setting our selves up for 2 years of whack-a-mole as we find and promote every case in the library (and some of those may not have as obvious resolutions).

Skimming the geopandas code it looks like passing `None` through for the keyword arguments will work as expected?

**Comment by TLouf** [comment]
> Skimming the geopandas code it looks like passing None through for the keyword arguments will work as expected?

As it builds one collection per geometry type, if, say, several lines are to be plotted and the user supplies an `edgecolor` as a list containing a None, it will raise.  Fixing it there would imply splitting in several collections whenever there is a `None` in a list-like kwarg. But well, the users can also substitute the `None` themselves with a default. It's just that at first I intuitively thought it would work, and it surprised me when I saw it did not.

But I understand all of the other points, that may be too much trouble for a not-so-great added value.
